File: php/src/FestivoClient.php
<?php
namespace Festivo;

use GuzzleHttp\Client as Guzzle;

/**
 * Festivo API client for accessing public holiday data.
 */
class FestivoClient {
    private $baseUrl;
    private $apiKey;
    private $http;

    /**
     * Initialize the Festivo API client.
     *
     * @param string|null $apiKey Your Festivo API key (or set FESTIVO_API_KEY env var)
     * @param string $baseUrl API base URL (default: https://api.getfestivo.com)
     */
    public function __construct($apiKey = null, $baseUrl = 'https://api.getfestivo.com') {
        $this->apiKey = $apiKey ?: getenv('FESTIVO_API_KEY');
        $this->baseUrl = $baseUrl;
        $this->http = new Guzzle(['base_uri' => $this->baseUrl]);
    }

    private function headers() {
        $h = ['Accept' => 'application/json'];
        if ($this->apiKey) {
            $h['Authorization'] = 'Bearer ' . $this->apiKey;
        }
        return $h;
    }

    private function request(string $path, array $params = []) {
        $res = $this->http->request('GET', $path, [
            'headers' => $this->headers(),
            'query' => array_filter($params, function($value) {
                return $value !== null;
            })
        ]);
        return json_decode($res->getBody()->getContents(), true);
    }

    /**
     * Get all holidays for a country and year.
     *
     * @param string $country ISO 3166-1 alpha-2 country code (e.g., "US", "GB", "IT")
     * @param int $year Year (e.g., 2026)
     * @param array $options Optional filters: regions, type, language, timezone
     * @return array Array with 'holidays' list and 'total' count
     */
    public function getHolidays(string $country, int $year, array $options = []) {
        $params = array_merge(['country' => $country, 'year' => $year], $options);
        return $this->request('/v3/public-holidays/list', $params);
    }

    /**
     * Get holidays for a specific city (Pro plan).
     *
     * @param string $country ISO 3166-1 alpha-2 country code (e.g., "IT")
     * @param string $cityCode City code in format {COUNTRY}-{CITY} (e.g., "IT-MILAN")
     * @param int $year Year (e.g., 2026)
     * @param array $options Optional filters: type, language, timezone
     * @return array Array with 'holidays' list and 'total' count
     */
    public function getCityHolidays(string $country, string $cityCode, int $year, array $options = []) {
        return $this->getHolidays($country, $year, array_merge(['regions' => $cityCode], $options));
    }

    /**
     * Get holidays for a specific region using ISO 3166-2 codes (Builder plan).
     *
     * @param string $country ISO 3166-1 alpha-2 country code (e.g., "GB")
     * @param string $regionCode ISO 3166-2 subdivision code (e.g., "GB-SCT" for Scotland)
     * @param int $year Year (e.g., 2026)
     * @param array $options Optional filters: type, language, timezone
     * @return array Array with 'holidays' list and 'total' count
     */
    public function getRegionalHolidays(string $country, string $regionCode, int $year, array $options = []) {
        return $this->getHolidays($country, $year, array_merge(['regions' => $regionCode], $options));
    }

    /**
     * Check if a specific date is a holiday.
     *
     * @param string $country ISO 3166-1 alpha-2 country code (e.g., "US")
     * @param string $date Date in format YYYY-MM-DD (e.g., "2026-12-25")
     * @param string|null $regions Optional comma-separated region/city codes
     * @return array Array with 'is_holiday' bool and optional 'holiday' array
     */
    public function checkHoliday(string $country, string $date, $regions = null) {
        $params = ['country' => $country, 'date' => $date];
        if ($regions !== null) {
            $params['regions'] = $regions;
        }
        return $this->request('/v3/public-holidays/check', $params);
    }
}

File: java/src/FestivoClient.java
package com.festivo;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Map;
import java.util.HashMap;
import org.json.JSONArray;
import org.json.JSONObject;

public class FestivoClient {
    private final String apiKey;
    private final String baseUrl;

    public FestivoClient(String apiKey) {
        this(apiKey, "https://api.getfestivo.com");
    }

    public FestivoClient(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
    }

    public HolidaysResponse getHolidays(String country, int year, Map<String, String> options) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("country", country);
        params.put("year", String.valueOf(year));
        if (options != null) params.putAll(options);
        JSONObject json = request("/v3/public-holidays/list", params);
        return HolidaysResponse.fromJson(json);
    }

    public HolidaysResponse getCityHolidays(String country, String cityCode, int year, Map<String, String> options) throws Exception {
        Map<String, String> opts = options == null ? new HashMap<>() : new HashMap<>(options);
        opts.put("regions", cityCode);
        return getHolidays(country, year, opts);
    }

    public HolidaysResponse getRegionalHolidays(String country, String regionCode, int year, Map<String, String> options) throws Exception {
        Map<String, String> opts = options == null ? new HashMap<>() : new HashMap<>(options);
        opts.put("regions", regionCode);
        return getHolidays(country, year, opts);
    }

    public JSONObject checkHoliday(String country, String date, String regions) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("country", country);
        params.put("date", date);
        if (regions != null) params.put("regions", regions);
        return request("/v3/public-holidays/check", params);
    }

    private JSONObject request(String path, Map<String, String> params) throws Exception {
        StringBuilder urlBuilder = new StringBuilder(baseUrl + path + "?");
        for (Map.Entry<String, String> entry : params.entrySet()) {
            urlBuilder.append(entry.getKey()).append("=").append(entry.getValue()).append("&");
        }
        URL url = new URL(urlBuilder.toString());
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Accept", "application/json");
        conn.setRequestProperty("Authorization", "Bearer " + apiKey);
        int status = conn.getResponseCode();
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        if (status != 200) throw new Exception("Festivo API error: " + status);
        return new JSONObject(response.toString());
    }

    // Minimal model classes
    public static class HolidaysResponse {
        public final JSONArray holidays;
        public HolidaysResponse(JSONArray holidays) {
            this.holidays = holidays;
        }
        public static HolidaysResponse fromJson(JSONObject json) {
            return new HolidaysResponse(json.getJSONArray("holidays"));
        }
    }
}
